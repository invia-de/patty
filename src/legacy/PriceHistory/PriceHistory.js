import React from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import UniversalPrice, {
  usePriceTotal
} from '../../components/atoms/UniversalPrice/UniversalPrice';
import NoBreak from '../../components/utilities/NoBreak/NoBreak';
import cx from '../../utils/classnames';
import {
  formatDate,
  DateTime
} from '../../components/utilities/DateTime/DateTime';
import Loading from '../../components/atoms/Loading/Loading';
import styles from './PriceHistory.module.scss';
import { ArrowLeft, ArrowRight } from '../../components/atoms/Icon/Icon';
import Tooltip from '../../components/atoms/Tooltip/Tooltip';
import travelService from '../../utils/travelService';
import url from '../../utils/url';
import isActive from '../../utils/features';
import noop from '../../utils/noop';

const ONE_DAY_IN_MILLISECONDS = 86400000;

/**
 * maps the URL value of the duration to a usable array of numbers
 * first index: duration to use within inDateRange check
 * second index: duration to use for calculation of returnDate when requesting data form the API
 */
const durationMap = {
  '-1': 14,
  '6_7': 7,
  '6_14': 14,
  '6_3-7': 7,
  '6_7-14': 14,
  '10': 8,
  '7': 12,
  '6_10': 10
};

class PriceHistory extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      /** complete pricechart data we hold - initial and further responses and possible placeholders */
      data: null,
      /** are we currently loading data from the API */
      loading: true,
      /** current position inside this.state.data from where we start the view */
      position: 0,
      view: [],
      /** for bar height transition */
      moved: true
    };

    this.oldHeights = {};

    // references for often used dates or timestamps
    this.today = formatDate(new Date(), 'yyyy-mm-dd')[0];
    this.tomorrow = new Date(this.today).getTime() + ONE_DAY_IN_MILLISECONDS;
    this.formattedTomorrow = formatDate(this.tomorrow, 'yyyy-mm-dd')[0];

    this.onClickPrev = this.onClickPrev.bind(this);
    this.onClickNext = this.onClickNext.bind(this);

    // params we need for the pricechart request
    this.params = {
      ...props.getParameters([
        'hotelId',
        'hotelIdType',
        'port',
        'adult',
        'area',
        'dest',
        'duration',
        'depDate',
        'retDate',
        'optPrice',
        'optOrganizer',
        'dynamicPricing',
        'ultSpecialTransfer',
        'depAirport',
        'optOcean',
        'optMeal',
        'roomtype',
        'topHotelSelected',
        'directFlight',
        'transferFilter',
        'children',
        'child1',
        'child2',
        'child3',
        'suppliers'
      ]),
      ...props.defaultParams
    };

    if (props.isFeatureActive('useAllBlockedOrganizerInIbe4', false)) {
      this.params.useAllBlockedOrganizer = 1;
    }

    if (props.isFeatureActive('blockOrganizerByTravelType', false)) {
      this.params.blockOrganizerByTravelType = 1;
    }

    // delete params that value is false
    for (let param in this.params) {
      if (this.params.hasOwnProperty(param)) {
        if (this.params[param] === false) {
          delete this.params[param];
        }
      }
    }

    this.retDate =
      this.params.retDate &&
      this.params.retDate
        .split('.')
        .reverse()
        .join('-');
    this.depDate =
      this.params.depDate &&
      this.params.depDate
        .split('.')
        .reverse()
        .join('-');

    if (this.retDate && this.depDate) {
      this.retDateTimestamp = new Date(this.retDate).getTime();
      this.depDateTimestamp = new Date(this.depDate).getTime();
      this.diffOfDateRange =
        (this.retDateTimestamp - this.depDateTimestamp) /
        ONE_DAY_IN_MILLISECONDS;

      if (this.params.duration && this.params.duration === '6_91') {
        this.params.duration = '6_' + this.diffOfDateRange;
      }
    }

    this.addDaysToRetDate = 14; // 14 is the fallback when we can not determine a number from the duration

    if (this.params.duration) {
      if (durationMap[this.params.duration]) {
        this.addDaysToRetDate = durationMap[this.params.duration];
      } else if (
        this.params.duration &&
        this.params.duration.indexOf('6_') === 0
      ) {
        this.addDaysToRetDate = parseInt(this.params.duration.substr(2));
      }
    }
  }

  inDateRange({ returnDate, departureDate }) {
    return this.retDate >= returnDate && this.depDate <= departureDate;
  }

  componentDidMount() {
    let newDepartureDate = this.params.depDate;
    let newReturnDate = this.params.retDate;

    if (this.diffOfDateRange < 14) {
      let add = Math.ceil(
        (14 - (this.diffOfDateRange - this.addDaysToRetDate)) / 2
      );

      newDepartureDate = formatDate(
        this.addDaysToDate(this.depDate, -add),
        'dd.mm.yyyy'
      )[0];

      newReturnDate = formatDate(
        this.addDaysToDate(this.retDate, add),
        'dd.mm.yyyy'
      )[0];
    } else {
      // always add additional days to the first request
      newReturnDate = formatDate(
        this.addDaysToDate(this.retDate, this.addDaysToRetDate),
        'dd.mm.yyyy'
      )[0];
    }

    this.props.getPricesFromAPI.get(
      {
        endpoint: 'search-pricechart',
        parameters: {
          ...this.params,
          depDate: newDepartureDate,
          retDate: newReturnDate
        }
      },
      result => {
        if (
          result.success &&
          result.response &&
          result.response.items &&
          result.response.items.length
        ) {
          let items = result.response.items;
          let position = 0;

          this.fillMissingDates(items);

          let inRangeCount = 0;
          let inRangeStartIndex = null;
          let minPrice = Infinity;
          let minPriceIndex = 0;

          items.forEach((obj, i) => {
            if (this.inDateRange(obj)) {
              inRangeCount++;
              if (inRangeStartIndex === null) {
                inRangeStartIndex = i;
              }
              if (minPrice > obj.priceInEuro) {
                minPrice = obj.priceInEuro;
                minPriceIndex = i;
              }
            }
          });

          let before = Math.floor((14 - inRangeCount) / 2);

          // if we have less than 14 inrange bars
          if (before >= 0) {
            position = inRangeStartIndex - before;

            if (position < 0) {
              position = 0;
            }
          } else {
            position = minPriceIndex - 6;

            if (position < 6) {
              position = 0;
            } else if (items.length - 1 - position < 14) {
              position = items.length - 14;
            }
          }

          let [view, fillCount] = this.getView(items, position);

          this.setState(
            {
              data: items,
              loading: position > items.length - 14 || position < 0,
              position: position,
              view: view,
              fillCount: fillCount
            },
            () => {
              this.getData();
            }
          );
        }
      }
    );
  }

  addFillerElement(arr, index, newDepDate) {
    arr.splice(index, 0, {
      placeholder: true,
      priceInEuro: null,
      loading: true,
      departureDate: newDepDate
    });
  }

  fillMissingDates(items) {
    for (let i = 0; i < items.length - 1; i++) {
      let diff =
        (new Date(items[i + 1].departureDate) -
          new Date(items[i].departureDate)) /
        ONE_DAY_IN_MILLISECONDS;
      if (diff !== 1) {
        this.addFillerElement(
          items,
          i + 1,
          formatDate(
            this.addDaysToDate(items[i].departureDate, 1),
            'yyyy-mm-dd'
          )[0]
        );
      }
    }
  }

  moveView(moveBy) {
    this.setState(
      ({ data, position }) => {
        let newPos = position + moveBy;
        const [view, fillCount] = this.getView(this.state.data, newPos);
        return {
          position: newPos,
          loading: newPos > data.length - 14 || newPos < 0,
          view: view,
          fillCount: fillCount,
          moved: true
        };
      },
      () => this.getData()
    );
  }

  onClickPrev() {
    this.moveView(-7);
  }

  onClickNext() {
    this.moveView(7);
  }

  mergeData(prevData, newData, oldView) {
    let result = {};
    let testSet = new Set();

    [...prevData, ...newData, ...oldView].forEach(function(obj) {
      if (testSet.has(obj.departureDate) && obj.priceInEuro === null) {
        return false;
      }

      if (obj.loading) {
        obj.placeholder = true;
      }

      result[obj.departureDate] = obj;
      testSet.add(obj.departureDate);
      return true;
    });

    return Object.values(result).sort(function(a, b) {
      if (a.departureDate > b.departureDate) return 1;
      if (a.departureDate < b.departureDate) return -1;
      return 0;
    });
  }

  getData() {
    if (this.state.loading) {
      let departureDate;
      let returnDate;

      this.state.view.forEach((item, i, arr) => {
        if (!departureDate && item.loading) {
          departureDate = formatDate(item.departureDate, 'dd.mm.yyyy')[0];
        } else if (departureDate && arr[i].loading) {
          returnDate = formatDate(
            this.addDaysToDate(item.departureDate, this.addDaysToRetDate),
            'dd.mm.yyyy'
          )[0];
        }
      });

      this.props.getPricesFromAPI.get(
        {
          endpoint: 'search-pricechart',
          parameters: {
            ...this.params,
            depDate: departureDate,
            retDate: returnDate
          }
        },
        result => {
          if (result.success && result.response && result.response.items) {
            this.fillMissingDates(result.response.items);

            this.setState(prevState => {
              let mergedData = this.mergeData(
                prevState.data,
                result.response.items,
                prevState.view
              );
              let newPos = this.state.position < 0 ? 0 : prevState.position;
              const [view, fillCount] = this.getView(mergedData, newPos);

              return {
                data: mergedData,
                view: view,
                fillCount: fillCount,
                loading: false,
                position: newPos
              };
            });
          }
        }
      );
    }
  }

  addDaysToDate(date, daysToAdd) {
    date = new Date(date);
    date.setDate(date.getDate() + daysToAdd);
    return date;
  }

  getFiller(fillCount, departureDate, dayModification) {
    let maxFill = 99;

    if (dayModification === -1) {
      const firstDate = departureDate;
      maxFill = (new Date(firstDate).getTime() - this.tomorrow) / 86400000;
    }

    if (fillCount > 14) {
      fillCount = 14;
    }

    if (fillCount > maxFill) {
      fillCount = maxFill;
    }

    let filler = new Array(fillCount).fill(1).map((_, i) => {
      return {
        loading: true,
        priceInEuro: null,
        departureDate: formatDate(
          this.addDaysToDate(departureDate, (i + 1) * dayModification),
          'yyyy-mm-dd'
        )[0]
      };
    });

    return dayModification === -1
      ? [filler.reverse(), fillCount]
      : [filler, fillCount];
  }

  getView(data, position) {
    if (position < 0) {
      let [filler, fillCount] = this.getFiller(
        ~position + 1,
        data[0].departureDate,
        -1
      );
      return [
        [...filler, ...(fillCount < 14 ? data.slice(0, 14 - fillCount) : [])],
        fillCount
      ];
    } else if (data.length < position + 14) {
      let [filler, fillCount] = this.getFiller(
        position + 14 - data.length,
        data[data.length - 1].departureDate,
        1
      );
      return [
        [...(position < data.length ? data.slice(position) : []), ...filler],
        fillCount
      ];
    }

    return [data.slice(position, 14 + position), 14, false];
  }

  render() {
    const { data, moved } = this.state;

    if (!data) {
      return null;
    }

    let view = this.state.view;
    const arr = view
      .map(obj => obj.priceInEuro)
      .filter(price => price !== null);
    const max = Math.max(...arr);
    const min = Math.min(...arr);
    const step = (max - min) / 100;

    const tempOldHeights = this.oldHeights;
    this.oldHeights = {};

    view = view.map(obj => {
      obj.cheapestInView = false;
      obj.inRange = false;

      const depDate = obj.departureDate;

      if (tempOldHeights[depDate]) {
        this.oldHeights[depDate] = tempOldHeights[depDate];
      }

      if (obj.loading) {
        obj.className = styles.bar_loading;
      } else if (obj.priceInEuro === min) {
        obj.className = styles.bar_cheapest;
        obj.cheapestInView = true;
        if (this.inDateRange(obj)) {
          obj.inRange = true;
        }
      } else if (!this.inDateRange(obj)) {
        obj.className = styles.bar_notInRange;
      } else {
        obj.className = styles.bar;
        obj.inRange = true;
      }

      return obj;
    });

    // trigger bar height transition
    if (moved) {
      setTimeout(() => {
        this.setState({ moved: false });
      }, 128);
    }

    return (
      <div className={styles.container}>
        <h2 className={cx('_styling-h2', styles.bold)}>Preisverlauf</h2>
        <div className={styles.chart}>
          {view.map(
            (
              {
                duration,
                className,
                loading,
                placeholder,
                price,
                priceTotal,
                priceInEuro,
                priceTotalInEuro,
                currency,
                departureDate
              },
              i,
              arr
            ) => {
              const newHeight = Math.floor(140 - (max - priceInEuro) / step);
              const renderResult = placeholder ? (
                <Tooltip
                  showArrow
                  classNameMessage={styles.tooltip}
                  key={departureDate}
                  message="Zu diesem Tag liegen uns leider keine Angebote vor."
                >
                  <div className={className} style={{ height: 31 }}>
                    <strong className={styles.price}>&nbsp;</strong>
                  </div>
                </Tooltip>
              ) : loading ? (
                <div key={departureDate} className={styles.loading_wrapper}>
                  <div className={className}>
                    <Loading />
                  </div>
                </div>
              ) : (
                <Tooltip
                  showArrow
                  onClick={event => {
                    event.persist();
                    this.props.onBarClick(event, arr[i]);
                  }}
                  classNameMessage={styles.tooltip}
                  key={departureDate}
                  message={
                    <div>
                      <NoBreak>
                        ab{' '}
                        <UniversalPrice
                          {...{
                            usePriceTotal: this.props.usePriceTotal,
                            price,
                            priceTotal,
                            priceInEuro,
                            priceTotalInEuro,
                            currency
                          }}
                        />
                        {this.props.usePriceTotal === false && ' p.P.'}
                      </NoBreak>
                      {`${duration} ${duration !== 1 ? 'Tage' : 'Tag'}`}
                    </div>
                  }
                >
                  <div
                    className={className}
                    style={{
                      height: moved
                        ? this.oldHeights[departureDate] || 31
                        : newHeight
                    }}
                  >
                    <strong className={styles.price}>
                      <UniversalPrice
                        {...{
                          usePriceTotal: this.props.usePriceTotal,
                          price,
                          priceTotal,
                          priceInEuro,
                          priceTotalInEuro,
                          currency
                        }}
                      />
                    </strong>
                    <div>
                      {`${duration} ${duration !== 1 ? 'Tage' : 'Tag'}`}
                    </div>
                  </div>
                </Tooltip>
              );

              this.oldHeights[departureDate] = placeholder ? 31 : newHeight;

              return renderResult;
            }
          )}
        </div>
        <div className={styles.axisContainer}>
          <button
            disabled={
              this.state.loading ||
              view[0].departureDate === this.formattedTomorrow
            }
            className={styles.button}
            onClick={this.onClickPrev}
          >
            <ArrowLeft />
          </button>
          <div className={styles.axis}>
            {view.map(obj => {
              return (
                <DateTime
                  className={styles.label}
                  value={obj.departureDate}
                  key={obj.departureDate}
                  format={
                    this.formattedTomorrow.substr(0, 4) !==
                    obj.departureDate.substr(0, 4)
                      ? 'wd dd.mm.yy'
                      : undefined
                  }
                />
              );
            })}
          </div>
          <button
            disabled={this.state.loading}
            className={styles.button}
            onClick={this.onClickNext}
          >
            <ArrowRight />
          </button>
        </div>
      </div>
    );
  }
}
PriceHistory.propTypes = {
  /** Function to run when user clicks a bar of the chart. It receives the object of that bar as second parameter */
  onBarClick: PropTypes.func,
  /** Function that must return a object of all the paramters and values you want to use for the travelservice request it takes a array of needed paramters as only parameter */
  getParameters: PropTypes.func,
  /** Function to check the availability of feautures on AIDU - must return a boolean */
  isFeatureActive: PropTypes.func,
  /** Performs the API request - will receive two parameters: first is a object of `{endpoint: string, parameters: object}` and e second parameter: callback that should run on complete of the request */
  getPricesFromAPI: PropTypes.shape({
    /**
     * { url:string, defaultParameters: object }
     * url: general url of the API endpoint in our case mobileapi.aidu.de/{endpoint}
     * defaultParameters: general parameters that do not change - e.g. agent
     */
    config: PropTypes.shape({
      url: PropTypes.string,
      defaultParameters: PropTypes.object
    }),
    get: PropTypes.func
  }),
  /** Wheter we should show total price or price per person */
  usePriceTotal: PropTypes.bool
};
PriceHistory.defaultProps = {
  onBarClick: (a, b) => console.log(a, b),
  getParameters: url.getAll,
  isFeatureActive: isActive,
  getPricesFromAPI: travelService,
  usePriceTotal: usePriceTotal
};

export default PriceHistory;

export function renderPriceHistory(props, container, callback) {
  ReactDOM.render(<PriceHistory {...props} />, container, callback);
}
