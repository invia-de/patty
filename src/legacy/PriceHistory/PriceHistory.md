```jsx
import travelService from '../../utils/travelService';
import response from './mock.json';

{
  travelService.mock('search-pricechart', false, response, 600);
}

<PriceHistory
  defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
  onBarClick={(event, data) => console.log('Bar clicked', data)}
/>;
```

You can use `folded` to make the chart foldable and folded by default:

```jsx
<PriceHistory
  defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
  onBarClick={(event, data) => console.log('Bar clicked', data)}
  folded
/>
```
