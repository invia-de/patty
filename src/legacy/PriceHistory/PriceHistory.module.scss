.container {
  font-family: sans-serif;
  font-size: 12px;
}

.chart,
.bar,
.axis {
  box-sizing: border-box;
}

.chart,
.axis {
  display: flex;
  align-items: flex-end;
  justify-content: center;
  flex-wrap: nowrap;
  width: calc(100% - 100px);
  margin: 0 auto;
  padding: 0 10px;
}

.chart {
  border-bottom: 1px solid #c6d8f5;
  height: 160px;
  padding: 10px 10px 0;

  > * {
    flex: 1 0 auto;
  }

  button {
    width: 100%;
    margin-bottom: -2px;
  }
}

.bar {
  font-size: 10px;
  box-sizing: border-box;
  flex: 1 1 auto;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  text-align: center;
  display: flex;
  padding: 5px 2px;
  background: #4284d2;
  color: #fff;
  transition: all 0.33s ease-out;
}

.bar_notInRange {
  composes: bar;
  background: #eaf2fe;
  color: #2a4d8f;
}

.bar_cheapest {
  composes: bar;
  background: #5fa800;
}

.label {
  composes: bar;
  font-size: 12px;
  text-align: center;
  height: auto;
  background: transparent;
  color: inherit;
  padding: 5px;
}

.axisContainer {
  display: flex;
  padding-top: 10px;
}

.button {
  line-height: 1;
  text-align: center;
  background: #2a4d8f;
  color: #fff;
  border: 0;
  margin: 0;
  padding: 5px;
  font-size: 2.2em;
  align-self: center;

  &:focus {
    outline: 2px solid #4284d2;
    outline-offset: 2px;
  }
}

.tooltip {
  text-align: center;

  strong {
    display: block;
    font-size: 16px;
    margin-bottom: 5px;
  }
}
