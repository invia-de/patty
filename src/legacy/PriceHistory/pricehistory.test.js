import React from 'react';
import PriceHistory from './PriceHistory';
import { render, getNodeText } from '@testing-library/react';
import travelService from '../../utils/travelService';
import response from './mock.json';

beforeEach(() => {
  travelService.mock('search-pricechart', true, response);
});

test('PriceHistory renders', () => {
  const { container } = render(
    <PriceHistory
      defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild.children[1].childNodes).toHaveLength(14);
});

test('PriceHistory mobile renders', () => {
  const { container } = render(
    <PriceHistory
      defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
      forceMobile
    />
  );
  expect(container.firstChild).toMatchSnapshot();
  expect(getNodeText(container.firstChild.children[1])).toEqual('Oktober 2019');
  expect(container.firstChild.children[2].childNodes).toHaveLength(7);
});

test('PriceHistory navigates to prev view', () => {
  const { container } = render(
    <PriceHistory
      defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
    />
  );
  container.firstChild.children[2].firstChild.click();
  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild.children[1].childNodes).toHaveLength(14);
});

test('PriceHistory navigates to next view', () => {
  const { container } = render(
    <PriceHistory
      defaultParams={{ depDate: '17.10.2019', retDate: '28.10.2019' }}
    />
  );
  container.firstChild.children[2].children[2].click();
  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild.children[1].childNodes).toHaveLength(14);
});

test('PriceHistory onBarClick hasBeenCalled', () => {
  const onClick = jest.fn();
  const { container } = render(<PriceHistory onBarClick={onClick} />);
  container.firstChild.children[1].firstChild.children[1].click();
  expect(onClick).toHaveBeenCalled();
});
