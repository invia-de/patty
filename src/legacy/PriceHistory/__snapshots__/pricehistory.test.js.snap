// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceHistory renders 1`] = `
<div
  class="container"
>
  <h2>
    Preisverlauf
  </h2>
  <div
    class="chart"
  >
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-0"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 351
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-0"
        class="tooltip__button"
        id="tooltip-button-0"
        type="button"
      >
        <div
          class="bar"
          style="height: 119.4392523364486px;"
        >
          <strong
            class="price"
          >
            € 351
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-1"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 350
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-1"
        class="tooltip__button"
        id="tooltip-button-1"
        type="button"
      >
        <div
          class="bar"
          style="height: 118.50467289719626px;"
        >
          <strong
            class="price"
          >
            € 350
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-2"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 356
               p.P.
            </strong>
            12
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-2"
        class="tooltip__button"
        id="tooltip-button-2"
        type="button"
      >
        <div
          class="bar"
          style="height: 124.11214953271028px;"
        >
          <strong
            class="price"
          >
            € 356
          </strong>
          <div>
            12
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-3"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 364
               p.P.
            </strong>
            11
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-3"
        class="tooltip__button"
        id="tooltip-button-3"
        type="button"
      >
        <div
          class="bar"
          style="height: 131.58878504672896px;"
        >
          <strong
            class="price"
          >
            € 364
          </strong>
          <div>
            11
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-4"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 341
               p.P.
            </strong>
            10
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-4"
        class="tooltip__button"
        id="tooltip-button-4"
        type="button"
      >
        <div
          class="bar"
          style="height: 110.09345794392523px;"
        >
          <strong
            class="price"
          >
            € 341
          </strong>
          <div>
            10
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-5"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 309
               p.P.
            </strong>
            9
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-5"
        class="tooltip__button"
        id="tooltip-button-5"
        type="button"
      >
        <div
          class="bar"
          style="height: 80.18691588785046px;"
        >
          <strong
            class="price"
          >
            € 309
          </strong>
          <div>
            9
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-6"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 281
               p.P.
            </strong>
            8
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-6"
        class="tooltip__button"
        id="tooltip-button-6"
        type="button"
      >
        <div
          class="bar"
          style="height: 54.018691588785046px;"
        >
          <strong
            class="price"
          >
            € 281
          </strong>
          <div>
            8
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-7"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 266
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-7"
        class="tooltip__button"
        id="tooltip-button-7"
        type="button"
      >
        <div
          class="bar_cheapest"
          style="height: 40px;"
        >
          <strong
            class="price"
          >
            € 266
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-8"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 266
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-8"
        class="tooltip__button"
        id="tooltip-button-8"
        type="button"
      >
        <div
          class="bar_cheapest"
          style="height: 40px;"
        >
          <strong
            class="price"
          >
            € 266
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-9"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 266
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-9"
        class="tooltip__button"
        id="tooltip-button-9"
        type="button"
      >
        <div
          class="bar_cheapest"
          style="height: 40px;"
        >
          <strong
            class="price"
          >
            € 266
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-10"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 266
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-10"
        class="tooltip__button"
        id="tooltip-button-10"
        type="button"
      >
        <div
          class="bar_cheapest"
          style="height: 40px;"
        >
          <strong
            class="price"
          >
            € 266
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-11"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 293
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-11"
        class="tooltip__button"
        id="tooltip-button-11"
        type="button"
      >
        <div
          class="bar"
          style="height: 65.23364485981308px;"
        >
          <strong
            class="price"
          >
            € 293
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-12"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 373
               p.P.
            </strong>
            8
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-12"
        class="tooltip__button"
        id="tooltip-button-12"
        type="button"
      >
        <div
          class="bar"
          style="height: 140px;"
        >
          <strong
            class="price"
          >
            € 373
          </strong>
          <div>
            8
             Tage
          </div>
        </div>
      </button>
    </div>
    <div
      class="tooltip"
    >
      <div
        class="tooltip__message"
        id="tooltip-13"
        role="tooltip"
      >
        <div
          class="tooltip"
        >
          <div
            class="nobreak"
          >
            <strong>
              ab 
              € 358
               p.P.
            </strong>
            7
             Tage,
            <br />
            ab 
            Berlin - Tegel
             (
            TXL
            )
          </div>
        </div>
      </div>
      <button
        aria-describedby="tooltip-13"
        class="tooltip__button"
        id="tooltip-button-13"
        type="button"
      >
        <div
          class="bar"
          style="height: 125.98130841121495px;"
        >
          <strong
            class="price"
          >
            € 358
          </strong>
          <div>
            7
             Tage
          </div>
        </div>
      </button>
    </div>
  </div>
  <div
    class="axisContainer"
  >
    <button
      class="button"
    >
      <svg
        aria-hidden="true"
        class="icon"
        viewBox="0 0 512 512"
      >
        <path
          d="M425 430L249 255 426 80 346 0 87 259l253 253 85-82z"
        />
      </svg>
    </button>
    <div
      class="axis"
    >
      <time
        class="label"
        datetime="2019-10-08"
      >
        Di 08.10.
      </time>
      <time
        class="label"
        datetime="2019-10-09"
      >
        Mi 09.10.
      </time>
      <time
        class="label"
        datetime="2019-10-10"
      >
        Do 10.10.
      </time>
      <time
        class="label"
        datetime="2019-10-11"
      >
        Fr 11.10.
      </time>
      <time
        class="label"
        datetime="2019-10-12"
      >
        Sa 12.10.
      </time>
      <time
        class="label"
        datetime="2019-10-13"
      >
        So 13.10.
      </time>
      <time
        class="label"
        datetime="2019-10-14"
      >
        Mo 14.10.
      </time>
      <time
        class="label"
        datetime="2019-10-15"
      >
        Di 15.10.
      </time>
      <time
        class="label"
        datetime="2019-10-16"
      >
        Mi 16.10.
      </time>
      <time
        class="label"
        datetime="2019-10-17"
      >
        Do 17.10.
      </time>
      <time
        class="label"
        datetime="2019-10-18"
      >
        Fr 18.10.
      </time>
      <time
        class="label"
        datetime="2019-10-19"
      >
        Sa 19.10.
      </time>
      <time
        class="label"
        datetime="2019-10-20"
      >
        So 20.10.
      </time>
      <time
        class="label"
        datetime="2019-10-21"
      >
        Mo 21.10.
      </time>
    </div>
    <button
      class="button"
    >
      <svg
        aria-hidden="true"
        class="icon"
        viewBox="0 0 512 512"
      >
        <path
          d="M88 82l176 176L87 432l80 80 259-259L173 0 88 82z"
        />
      </svg>
    </button>
  </div>
</div>
`;
